File: Sorting/Insertion Sort/Python/InsertionSort.py
def insertionSort(array):
    if array == None or len(array) == 0:
        return None
    for i in range(1,len(array)):
        currentElement = array[i]
        j = i - 1
        while j >= 0:
            if array[j] > currentElement:
                array[j+1] = array[j]
            else:
                break
            j-=1
        array[j+1] = currentElement
    return array
#
# Sample Input: [4,3,2,1]
# Sample Output: [1,2,3,4]
#
# Time Complexity in Best Case: O(n)
# Time Complexity in Average Case: O(n^2)
# Time Complexity in Worst Case: O(n^2)
# Space Complexity in Worst Case: O(1)
#

File: Sorting/Insertion Sort/Java/InsertionSort.java

public class InsertionSort {
	
	public static int[] insertionSort(int[] array) {
		if (array == null || array.length == 0) //if array is not sortable, then return null
			return null;
		
		for (int i = 1; i < array.length; i++) {
			int currentElement = array[i];
			int j = i-1;
			for (;j >= 0; j--) //we move the current element to the correct position at the left of the array in this step
				if (array[j] > currentElement)
					array[j+1] = array[j];
				else 
					break;
				
			array[j+1] = currentElement; //we placed the current element in correct position
		}		
		return array;
	}
	
}
	
	/*
	 * Sample Input: [4,3,2,1]
	 * Sample Output: [1,2,3,4]
	 * 
	 * Time Complexity in Worst Case: O(n^2)
	 * Space Complexity in Worst Case: O(1)
	 * 
	 */


